app/controlplane/internal/usercontext/jwt_middleware.go

```diff
@@ -0,0 +1,231 @@
+	ErrTokenInvalid           = errorsAPI.Unauthorized(reason, "Token is invalid")
+	ErrTokenExpired           = errorsAPI.Unauthorized(reason, "JWT token has expired")
+	ErrTokenParseFail         = errorsAPI.Unauthorized(reason, "Fail to parse JWT token ")
+	ErrUnSupportSigningMethod = errorsAPI.Unauthorized(reason, "Wrong signing method")
+	ErrWrongContext           = errorsAPI.Unauthorized(reason, "Wrong context for middleware")
+	ErrNeedTokenProvider      = errorsAPI.Unauthorized(reason, "Token provider is missing")
+)
+
+type JWTAuthContext struct {
+	Claims         jwt.Claims
+	IsAPIToken     bool
+	IsRobotAccount bool
+}
+
+type authzContextKey struct{}
+
+type JWTOption func(*options)
+type TokenProviderOption func(*providerOption)
+
+type VerifyAudienceFunc func(*jwt.Token) bool
+
+type providerOption struct {
+	providerKey        string
+	signingMethod      jwt.SigningMethod
+	keyFunc            jwt.Keyfunc
+	claims             func() jwt.Claims
+	verifyAudienceFunc VerifyAudienceFunc
+}
+
+func WithSigningMethod(method jwt.SigningMethod) TokenProviderOption {
+	return func(o *providerOption) {
+		o.signingMethod = method
+	}
+}
+
+func WithProviderKey(providerKey string) TokenProviderOption {
+	return func(o *providerOption) {
+		o.providerKey = providerKey
+	}
+}
+
+func WithClaims(f func() jwt.Claims) TokenProviderOption {
+	return func(o *providerOption) {
+		o.claims = f
+	}
+}
+
+func WithKeyFunc(keyFunc jwt.Keyfunc) TokenProviderOption {
+	return func(o *providerOption) {
+		o.keyFunc = keyFunc
+	}
+}
+
+func WithVerifyAudienceFunc(f VerifyAudienceFunc) TokenProviderOption {
+	return func(o *providerOption) {
+		o.verifyAudienceFunc = f
+	}
+}
+
+type options struct {
+	tokenProviders []providerOption
+}
+
+func WithTokenProvider(opts ...TokenProviderOption) JWTOption {
+	op := &providerOption{}
+	for _, opt := range opts {
+		opt(op)
+	}
+	return func(o *options) {
+		o.tokenProviders = append(o.tokenProviders, *op)
+	}
+}
+
+func WithJWTTokenMiddleware(opts ...JWTOption) middleware.Middleware {
+	o := &options{}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return func(handler middleware.Handler) middleware.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			if header, ok := transport.FromServerContext(ctx); ok {
+				auths := strings.SplitN(header.RequestHeader().Get(authorizationKey), " ", 2)
+				if len(auths) != 2 || !strings.EqualFold(auths[0], bearerWord) {
+					return nil, ErrMissingJwtToken
+				}
+				jwtToken := auths[1]
+				var (
+					tokenInfo *jwt.Token
+					err       error
+				)
+
+				tokenProviderLen := len(o.tokenProviders)
+				for idx, provider := range o.tokenProviders {
+					tokenInfo, err = runProviderValidator(provider, jwtToken)
+
+					// Check if it's the last provider and still failed
+					if err != nil {
+						if idx != tokenProviderLen-1 {
+							continue
+						}
+
+						return nil, err
```

Comment:
I tend to wrap errors to add tracing, I'd suggest you do the same `fmt.Errorf("... %w", err)`